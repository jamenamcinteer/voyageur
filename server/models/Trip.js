const mongoose = require("mongoose");
const { Schema } = mongoose;

const tripSchema = new Schema({
  destination: String,
  endDate: Number,
  startDate: Number,
  photo: String,
  photoAttribution: String,
  uid: String
});

mongoose.model("trips", tripSchema);
