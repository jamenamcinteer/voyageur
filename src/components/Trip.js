import React, { useState } from "react";
import TripHeader from "./Navigation/TripHeader";
import ButtonLink from "./Buttons/ButtonLink";
import Button from "./Buttons/Button";
import BudgetCard from "./Cards/BudgetCard";
import { startRemoveTrip } from "../actions/trips";
import { startRemoveBudgetCategory } from "../actions/budgetCategories";
import { startRemoveBudgetItem } from "../actions/budgetItems";
import { startRemoveExpense } from "../actions/expenses";
import useSortAscendingAlphabetical from "../hooks/useSortAscendingAlphabetical";
import Modal from "react-modal";
import { ModalText } from "./StyledComponents/Modals";
import { Container, CardContainer } from "./StyledComponents/Layout";
import { MainButtons } from "./StyledComponents/Forms";
import { connect } from "react-redux";

const Trip = props => {
  const [deleteModal, setDeleteModal] = useState(false);

  const trip = props.trips.find(trip => trip._id === props.match.params.id);
  const budgetCategories = useSortAscendingAlphabetical(
    props.budgetCategories.filter(bCategory => bCategory.tripId === trip._id),
    "budgetCategory"
  );

  const deleteTrip = () => {
    startRemoveTrip(trip._id);

    props.budgetCategories.map(i => {
      if (i.tripId === trip._id) startRemoveBudgetCategory(i._id);
      return true;
    });

    props.budgetItems.map(i => {
      if (i.tripId === trip._id) startRemoveBudgetItem(i._id);
      return true;
    });

    props.expenses.map(i => {
      if (i.tripId === trip._id) startRemoveExpense(i._id);
      return true;
    });

    props.history.push("/");
  };

  return (
    <React.Fragment>
      {trip && (
        <React.Fragment>
          <TripHeader
            title="Trip Details"
            theme={props.theme}
            backTo="/"
            trip={trip}
            budgetItems={props.budgetItems}
            expenses={props.expenses}
            auth={props.auth}
          />
          {props.budgetCategories.length > 0 && props.budgetItems.length > 0 && (
            <Container style={{ ...{ textAlign: "center" } }}>
              <ButtonLink
                to={`/trip/${trip._id}/add-expense`}
                buttonText="Add Expense"
                buttonType="primary"
                theme={props.theme}
                customStyles={{
                  background: {
                    display: "inline-block",
                    width: "auto"
                  }
                }}
              />
            </Container>
          )}
          <CardContainer>
            {budgetCategories.length > 0 &&
              budgetCategories.map(budgetCategory => {
                return (
                  <BudgetCard
                    key={budgetCategory._id}
                    actual={2130}
                    budgeted={3760}
                    theme={props.theme}
                    budgetCategory={budgetCategory}
                    budgetItems={props.budgetItems}
                    expenses={props.expenses}
                  />
                );
              })}
          </CardContainer>
          <Container style={{ ...{ textAlign: "center" } }}>
            <ButtonLink
              to={`/trip/${trip._id}/add-budget-category`}
              buttonText="Add Budget Category"
              buttonType={
                props.budgetCategories.length > 0 ? "secondary" : "primary"
              }
              theme={props.theme}
              customStyles={{
                background: {
                  display: "inline-block",
                  width: "auto"
                }
              }}
            />
            <br />
            <ButtonLink
              to={`/trip/${trip._id}/edit`}
              buttonText="Edit Trip"
              buttonType="link"
              theme={props.theme}
              customStyles={{
                background: {
                  display: "inline-block",
                  width: "auto"
                }
              }}
            />
            <br />
            <Button
              handleClick={e => setDeleteModal(true)}
              buttonText="Delete Trip"
              buttonType="delete"
              theme={props.theme}
              customStyles={{
                background: {
                  display: "inline-block",
                  width: "auto"
                }
              }}
            />
            <Modal
              isOpen={deleteModal}
              onRequestClose={e => setDeleteModal(false)}
              contentLabel="Delete Modal"
              style={{
                overlay: {
                  backgroundColor: "rgba(0,0,0,.5)"
                },
                content: {
                  borderRadius: "0",
                  minHeight: "20vh",
                  maxHeight: "40vh",
                  overflowY: "auto",
                  width: "80vw",
                  top: "30vh",
                  left: "10vw",
                  right: "auto",
                  bottom: "auto"
                }
              }}
            >
              <ModalText>
                Are you sure you want to permanently delete this trip and all
                its budget categories, budget items, and expenses?
              </ModalText>
              <MainButtons>
                <Button
                  theme={props.theme}
                  handleClick={e => setDeleteModal(false)}
                  buttonText="Cancel"
                  buttonWidth="auto"
                  buttonType="link"
                  buttonDisplay="inline"
                  customStyles={{ background: { padding: "10px 0" } }}
                />
                <Button
                  theme={props.theme}
                  handleClick={deleteTrip}
                  buttonText="Yes, Delete"
                  buttonWidth="auto"
                  buttonType="primary"
                  buttonDisplay="inline"
                />
              </MainButtons>
            </Modal>
          </Container>
        </React.Fragment>
      )}
    </React.Fragment>
  );
};

const mapStateToProps = state => {
  return {
    trips: state.trips,
    budgetCategories: state.budgetCategories,
    budgetItems: state.budgetItems,
    expenses: state.expenses,
    auth: state.auth
  };
};

export default connect(mapStateToProps)(Trip);
