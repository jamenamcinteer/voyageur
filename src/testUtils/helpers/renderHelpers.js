import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../../theme";
import "jest-styled-components";
import { createStore } from "redux";
import { Provider } from "react-redux";
import tripsReducer, { tripsReducerDefaultState } from "../../reducers/trips";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

export function renderWithTheme(component) {
  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
}

export function renderWithReduxRouterAndTheme(
  ui,
  {
    initialState,
    store = createStore(tripsReducer, tripsReducerDefaultState),
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <Router history={history}>{ui}</Router>
        </ThemeProvider>
      </Provider>
    ),
    store,
    history
  };
}
