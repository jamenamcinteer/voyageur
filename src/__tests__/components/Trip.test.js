import React from "react";
import { cleanup } from "@testing-library/react";
import { createStore } from "redux";
import Trip from "../../components/Trip";
import {
  trips,
  budgetCategories,
  budgetItems,
  expenses,
  auth
} from "../../testUtils/fixtures/mockStoreData";
import { renderWithReduxRouterAndTheme } from "../../testUtils/helpers/renderHelpers";

// function renderWithRedux(
//   ui,
//   {
//     initialState,
//     store = createStore(tripsReducer, tripsReducerDefaultState)
//   } = {}
// ) {
//   return {
//     ...render(<Provider store={store}>{ui}</Provider>),
//     store
//   };
// }

// function renderWithRouter(
//   ui,
//   {
//     route = "/",
//     history = createMemoryHistory({ initialEntries: [route] })
//   } = {}
// ) {
//   return {
//     ...render(<Router history={history}>{ui}</Router>),
//     history
//   };
// }

afterEach(cleanup);

// it("renders", () => {
//   const { asFragment } = render(
//     <Provider store={store}>
//       <Trip trips={trips} />
//     </Provider>
//   );
//   expect(asFragment()).toMatchSnapshot();
// });

// test("can render with redux with defaults", () => {
//   // const { getByTestId, getByText } = renderWithRedux(<Trip />)
//   // fireEvent.click(getByText('+'))
//   // expect(getByTestId('count-value').textContent).toBe('1')
//   const { asFragment } = renderWithRedux(<Trip />);
//   expect(asFragment()).toMatchSnapshot();
// });

test("can render with redux with custom store", () => {
  // this is a silly store that can never be changed
  const store = createStore(() => ({
    trips,
    budgetCategories,
    budgetItems,
    expenses,
    auth
  }));
  const { asFragment } = renderWithReduxRouterAndTheme(
    <Trip match={{ params: { id: "1" } }} />,
    {
      store
    }
  );
  // const { getByTestId, getByText } = renderWithRedux(<Trip />, {
  //   store,
  // })
  // fireEvent.click(getByText('+'))
  // expect(getByTestId('count-value').textContent).toBe('1000')
  // fireEvent.click(getByText('-'))
  // expect(getByTestId('count-value').textContent).toBe('1000')
  expect(asFragment()).toMatchSnapshot();
});
