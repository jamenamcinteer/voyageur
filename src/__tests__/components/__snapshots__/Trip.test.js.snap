// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render with redux with custom store 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <div>
          <div
            class="sc-bdVaJa hZqfbl"
          >
            <a
              href="/"
            >
              <svg
                aria-label="Go Back"
                style="fill: #4b5c6b; width: 20px; height: 20px; float: left; margin-top: 2px;"
                version="1.1"
                viewBox="0 0 199.404 199.404"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <polygon
                    points="135.412,0 35.709,99.702 135.412,199.404 163.695,171.119 92.277,99.702 163.695,28.285 	"
                  />
                </g>
              </svg>
            </a>
            <h1
              class="sc-bwzfXH OVfcq"
            >
              Trip Details
            </h1>
            <button
              class="sc-htpNat fqCEYg"
            >
              <img
                alt="Account Options"
                class="sc-bxivhb jwtarI"
                src="https://picsum.photos/id/947/45/45"
              />
            </button>
          </div>
          <div
            style="position: relative;"
          >
            <div
              style="background-image: url(https://images.unsplash.com/photo-1549317487-92cec5e76645?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjc5ODc0fQ&w=300&h=100&fit=crop&crop=focalpoint); background-repeat: no-repeat; background-size: cover; filter: brightness(0.6); height: 150px;"
            />
            <div
              style="height: 150px; padding: 0px 20px; color: rgb(255, 255, 255); font-size: 1em; font-weight: normal; position: absolute; width: 100%; top: 0px;"
            >
              <div
                style="display: flex; justify-content: space-between; align-items: center; height: 90px;"
              >
                <h2>
                  Hawaii
                </h2>
                <h3>
                  Apr 26, 1970 - Oct 1, 8307
                </h3>
              </div>
              <div
                style="width: 100%; height: 25px; position: relative; background: rgb(220, 224, 227); border-radius: 5px;"
              >
                <span
                  style="display: block; height: 100%; border-radius: 5px; position: relative; overflow: hidden; width: 29.731%; background-color: rgb(12, 124, 136);"
                />
                <div
                  style="position: absolute; top: 0px; color: white; left: 10px; line-height: 25px;"
                >
                  $595 of $2000
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-fjdhpX cEgVjZ"
        style="text-align: center;"
      >
        <a
          border="0"
          buttontype="primary"
          class="sc-bZQynM emmYYP"
          href="/trip/1/add-expense"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-gzVnrw dfkowZ"
            color="#fff"
          >
            Add Expense
          </span>
        </a>
      </div>
      <div
        class="sc-jzJRlG kQMCmT"
      >
        <div
          class="sc-gZMcBi ibexuG"
        >
          <h3
            class="sc-gqjmRU cBbVf"
          >
            <span>
              Airfare
            </span>
            <a
              backgroundcolor="transparent"
              border="0"
              buttontype="link"
              class="sc-bZQynM iSrURx"
              href="/trip/1/budget-category/1/edit"
              style="padding: 0px 10px; margin: 0px; width: auto; font-size: .8em;"
            >
              <span
                class="sc-gzVnrw ejnAGc"
              >
                Edit
              </span>
            </a>
          </h3>
          <div
            style="width: 100%; height: 25px; position: relative; background: rgb(220, 224, 227); border-radius: 5px;"
          >
            <span
              style="display: block; height: 100%; border-radius: 5px; position: relative; overflow: hidden; width: 29.731%; background-color: rgb(12, 124, 136);"
            />
            <div
              style="position: absolute; top: 0px; color: white; left: 10px; line-height: 25px;"
            >
              $595 of $2000
            </div>
          </div>
          <hr
            class="sc-VigVT fgNDYA"
          />
          <div>
            <h3
              class="sc-iwsKbI cEoBnw"
            >
              <span>
                Flights
              </span>
              <a
                backgroundcolor="transparent"
                border="0"
                buttontype="link"
                class="sc-bZQynM iSrURx"
                href="/trip/1/budget-category/1/budget-item/1"
                style="padding: 0px 10px; margin: 0px; width: auto; font-size: .8em;"
              >
                <span
                  class="sc-gzVnrw ejnAGc"
                >
                  View Expenses
                </span>
              </a>
              <a
                backgroundcolor="transparent"
                border="0"
                buttontype="link"
                class="sc-bZQynM iSrURx"
                href="/trip/1/budget-category/1/budget-item/1/edit"
                style="padding: 0px 10px; margin: 0px; width: auto; font-size: .8em;"
              >
                <span
                  class="sc-gzVnrw ejnAGc"
                >
                  Edit
                </span>
              </a>
            </h3>
            <div
              style="width: 100%; height: 25px; position: relative; background: rgb(220, 224, 227); border-radius: 5px;"
            >
              <span
                style="display: block; height: 100%; border-radius: 5px; position: relative; overflow: hidden; width: 29.731%; background-color: rgb(12, 124, 136);"
              />
              <div
                style="position: absolute; top: 0px; color: white; left: 10px; line-height: 25px;"
              >
                $595 of $2000
              </div>
            </div>
          </div>
          <div
            style="text-align: center;"
          >
            <a
              backgroundcolor="#fff"
              border="1px solid undefined"
              buttontype="secondary"
              class="sc-bZQynM epODmH"
              href="/trip/1/budget-category/1/add-budget-item"
              style="display: inline-block; width: auto; margin: 20px 0px 0px 0px;"
            >
              <span
                class="sc-gzVnrw hcLogF"
              >
                Add Budget Item
              </span>
            </a>
          </div>
        </div>
      </div>
      <div
        class="sc-fjdhpX cEgVjZ"
        style="text-align: center;"
      >
        <a
          backgroundcolor="#fff"
          border="1px solid undefined"
          buttontype="secondary"
          class="sc-bZQynM epODmH"
          href="/trip/1/add-budget-category"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-gzVnrw hcLogF"
          >
            Add Budget Category
          </span>
        </a>
        <br />
        <a
          backgroundcolor="transparent"
          border="0"
          buttontype="link"
          class="sc-bZQynM iSrURx"
          href="/trip/1/edit"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-gzVnrw ejnAGc"
          >
            Edit Trip
          </span>
        </a>
        <br />
        <button
          class="sc-htoDjs cawPai"
          style="display: inline-block; width: auto;"
        >
          <span
            class="sc-dnqmqq gdGOzX"
          >
            Delete Trip
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
